### ISSUE TITLE
Planned puzzles masterlist

### ISSUE THREAD

**Issue by HactarCE** [HactarCE/Hyperspeedcube#55]
If you would like to see a puzzle added that is not listed here, add a comment on this issue.

At this time, I am only accepting puzzles that are ≤5D and can be built and simulated using current features. That means: 

- Euclidean N-dimensional space
- only flat cuts
- no deleted pieces 
- only doctrinaire turns
- no multi-core

I will check off puzzles in this list once their production-ready implementation is merged into the `v2.0` branch.

# Must-have for v2.0

This is the list of puzzles that I know people will want in v2.0 and have minimal complications. Namely: they are all doctrinaire, have few enough axes that we can reasonably name all the twists, and exist in 3D or 4D.

## 3D

- [x] N^3 [Rubik's Cube](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=2968) generator
- [ ] AxBxC cuboid generator (needs piece types)
- [ ] N-gon face-turning prism generator (needs piece types)
- [ ] N-layer [Skewb](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=621) generator (needs piece types refactor)
- [x] N-layer [Megaminx](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=650) generator
- [x] N-layer Kilominx generator
- [ ] N-layer [Pyraminx](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=539) generator
- [ ] N-layer [Halpern-Meier Tetrahedron](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=547) generator
- [ ] N-layer [Face-Turning Octahedron](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=1663) generator (needs piece types refactor)
- [x] [Dino Cube](https://www.twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=5020)
- [x] [Compy Cube](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=1821)

## 4D

- [x] N^4 Rubik's Hypercube generator
- [ ] AxBxCxD hypercuboid generator (needs piece types)
- [ ] {A}x{B} duoprism generator (needs piece types)
- [ ] N-layer simplex generator (at various cut depths)
- [ ] 120-cell (needs facet/axis names and piece types)

# Shortlist

These puzzles will be possible to implement in v2.0 but have lower priority. They will be added as they are implemented.

## 3D

- [ ] Complex 3x3x3 (partially implemented)
- [x] Deep-cut face-turning dodecahedra
    - [x] Megaminx-Crystal Intermediate
    - [x] [Pyraminx Crystal](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=652)
    - [x] [Starminx](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=1759)
    - [ ] [Curvy Starminx](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=4344) (missing generator + piece types + tags)
    - [ ] N-layer [Pentultimate](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=1741) generator (needs piece types refactor)
- [ ] Vertex-turning icosahedra
    - [ ] N-layer [Tutt's Icosaminx](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=1640) generator
    - [ ] N-layer [Icosamate](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=1743) generator
        - [ ] [Regular Astrominx](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=6677)
- [ ] [Mixup family](https://github.com/user-attachments/assets/5d957f13-c942-4367-89df-7e4d86b0da70)
  - [ ] Mixup Cube
  - [ ] Bagua Cube (needs layer twist restrictions)
- [x] [Dino Skewb](https://twistypuzzles.com/app/museum/museum_showitem.php?pkey=1618)
- [ ] [Dino Dodecahedron](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=1760) (also called [Rediminx](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=7436))
- [ ] N-layer [Tuttminx](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=1557) generator
- [ ] N-layer [Trajber's Octahedron](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=401) (CTO) generator
- [ ] N-layer Comboctahedron (not a generator yet, needs piece types, needs axis names)
- [ ] N-layer [(name TBD) Vertex-Turning Octahedron](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=1923)
- [ ] N-layer [Dino Octa](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=2673) generator
- [ ] [F-Skewb](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=1723)
- [ ] FTCO (needs Luna to make it a generator)

## 4D

- [ ] Megaminx prism generator (needs piece types)
- [ ] Facet-Turning 24-cell (needs piece types)
- [ ] N-layer Trajber's 16-cell (vertex-turning; shapemod of 3^4) generator
- [ ] Ridge-Turning Hypercube
- [ ] Demi N^4 generator
- [ ] Facet-Turning 120-cell

## 5D

These are waiting on #54

- [ ] N^5 Rubik's Hypercube
- [ ] AxBxCxDxE hypercuboid generator

**Comment by milojacquet** [comment]
Deeper face-turning dodecahedra: pyraminx crystal, (curvy) starminx, (master) pentultimate

**Comment by milojacquet** [comment]
N-layer skewb generator

**Comment by OtaMota** [comment]
"Trojber's" 16 cell (3^4 shape mod, vertex turning 16 cell)

**Comment by mniip** [comment]
120-cell seems like a gaping omission. Or are its 120 grips too much for "few enough axes that we can reasonably name all the twists"?

**Comment by OtaMota** [comment]
Hypercuboids are omitted for 4d for some reason when they are listed for 3 and 5d

**Comment by HactarCE** [comment]
Good point! I've added FT 120 to the shortlist

**Comment by milojacquet** [comment]
N-layer icosamate/astrominx generator
Face**t**-turning 120-cell

**Comment by cryofractal** [comment]
If we have FTO we might as well have cubeoctahedron as well (but that may be slightly too obscure?)

**Comment by milojacquet** [comment]
I'm not sure it needs an Astrominx generator in that sense. What I meant was the mf8 Astrominx, also called [Regular Astrominx](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=6677), which is the 3-layer version of the Icosamate.

**Comment by milojacquet** [comment]
Tuttminx generator

**Comment by cryofractal** [comment]
Facet-turning 24 cell for the record

**Comment by vinfluorine** [comment]
antiprism generator

**Comment by milojacquet** [comment]
[F-skewb](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=1723) (could be made N-layer, as far I know that has never been done)
Cheese (prisms where the equatorial axes' cuts pass through the origin)
N-layer [This octahedron](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=1923) generator
[Rediminx](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=7436)/[Dino Dodecahedron](https://twistypuzzles.com/cgi-bin/puzzle.cgi?pkey=1760) (also known as Starminx I but sometimes Starminx I is the face-turning one)

**Comment by HactarCE** [comment]
Postponing Cheese because we want a general cheese generator, and some cheeses jumble. (e.g., octagonal 2-layer cheese has 2^3 + 45-degree turns)

**Comment by HactarCE** [comment]
https://alpha.twizzle.net/explore/?puzzle-description=d+v+0.74 AJ Bauhinia II

**Comment by OtaMota** [comment]
https://twistypuzzles.com/app/museum/museum_showitem.php?pkey=3784 https://youtu.be/gZRgOm0U3Wg?si=MHTzjssNbvCZTEl2 Generator (tho as previously discussed, I am aware that puzzles like this would not be feasible in 2.0)

**Comment by HactarCE** [comment]
I've added checkboxes for the tasks remaining for all the must-have puzzles & generators. Notation, twists directions, and metadata & tags require more HSC features, but the rest can be checked off before then. Once a puzzle is totally done, its version in the Lua file should be set to `1.0.0`.

**Comment by NonerKao** [comment]
Hi,

Any progress on `Facet-Turning 24-cell`? I should put it this way: **I would like to help to build 24-cell puzzles**, but I wonder if you already have any plans on the implementation, or any specific directions. If so, then I would be happy to follow them. 

For now, I don't have a clear picture on how to do it, but I have read through [wikipedia: 24-cell](https://en.wikipedia.org/wiki/24-cell) and [24-cell related threads in MC4D Archive](https://superliminal.com/cube/MessageArchive.html), so at least the geometry is not alien to me.

Finally, some background of me:

* [A simple PR](https://github.com/HactarCE/Hyperspeedcube/pull/15) to this software was accepted.
* I got interest in hypercubing (but not yet speed solving, not sure if I will ever get into that realm), so I started [a project](https://ithelp.ithome.com.tw/users/20103524/ironman/5183) back in 2022/Q3, in which I
   * Solved 2x2x2x2 using MC4D
   * [Created 4D skewb (vertex-turned hypercube) from pure web GL](https://github.com/NonerKao/webgl-ironman), and solved it.
 * I wasn't familiar with Rust when I sent the PR, but I just finished [a project that implements a boardgame](https://github.com/NonerKao/PathogenEngine), so it is not weird to me now. Still, this project uses some features that are beyond my current knowledge, but I think they will be understandable sooner or later.

**Comment by HactarCE** [comment]
@NonerKao I'm glad you'd like to contribute! I have excellent news: the FT 24-cell is already implemented, and you can play with it on the v2.0 branch, although there's no way to scramble or save progress so I wouldn't recommend solving it.

All the puzzles are written in [Lua](https://lua.org/). The API documentation is [here](https://dev.hypercubing.xyz/hsc/lua/basics/), and should be very nearly up-to-date. The FT 24-cell is already implemented in [this file](https://github.com/HactarCE/Hyperspeedcube/blob/v2.0/lua/puzzles/ft_24_cell.lua). Here's what's left to be done:
1. Name for each cell. The code can be generated using the built-in "Puzzles → Dev Tools → Lua Generator" then "Load Orbit" and select the orbit of 24 colors, then rename and reorder them. Ideally we want prefix-free names (e.g., if there is a color named `FR`, then there should not be a color named `F`).
2. Color system with default colors. See `lua/symmetries/hypercubic.lua` for an example. Relevant docs are [here](https://dev.hypercubing.xyz/hsc/lua/color-system-library/) and [here](https://dev.hypercubing.xyz/hsc/lua/geometry/orbit/#orbitnamed). The existing puzzle already has some colors assigned, but it's using an older format that will break when reordering cells.
3. Piece type names. There is a standard for these names that @milojacquet has developed, but even just isolating each piece type would be helpful. I plan to add tools in the program for identifying piece types, but for now you have to do it manually. See `ft_cubes.lua` for an example.
4. Tags (properties like "cell-turning," "shallow-cut," "24-cell," "doctrinaire"). I'm currently implement this in Rust and expect it to be done within a week or two.
5. Twist notation. This is not yet implemented in Rust, but I will do it soon.

1, 2, and 3 are things you can do now if you would like to. 4 and 5 are things I will do as part of updating all the puzzle files once the Rust side is complete. Piece types in particular are something that almost all of the current puzzles are missing, and would be very helpful. The relevant docs are [here](https://dev.hypercubing.xyz/hsc/lua/puzzle-construction/puzzle/#puzzlemark_piece).

Most of the work at this stage in development is in Rust, and I have strong feelings about how it should be implemented. The good news is that individual puzzles are very straightforward to implement, and I plan to add more tools to make it as easy as possible. The hardest part is naming the colors, twist axes, and piece types.

**Comment by HactarCE** [comment]
Closing this issue in favor of #74
